[PM-20344] Export: Attachments with the same name are not included/overwritten

djsmith85 opened the pull request:
## 🎟️ Tracking
<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->
https://bitwarden.atlassian.net/browse/PM-20344

## 📔 Objective
<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->
In case a cipher has 2 attachments with the same filename, they were overwritten within the zip export

~~This appends an underscore as a separator and the attachmentId to truly distinguish the files by name~~
This create a folder for each attachment and uses the `attachmentId` as the folder-name. The previous solution did not take file extensions into account, which occurred to me late last night. The solution proposed by @kpiris down below, is nice and simple. It also has the added benefit of retaining the attachmentId which might be handy when restoring a backup.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

djsmith85 committed:
Fix exporting cipher attachments with the same filename

In case a cipher has 2 attachments with the same filename, they were overwritten within the zip export

This appends an underscore as a separator and the attachmentId to truly distinguish the files by name

kpiris commented:
> This appends an underscore as a separator and the attachmentId to truly distinguish the files by name

I haven't reviewed the code, but according to this description I'm not a fan of this way of avoiding overwriting them.

I recall the structure of the attachments in the zip file was something like this:

```
.
├── itemId01
│   └── attachmentFile
├── itemId02
│   └── attachmentFile
└── itemId03
```

What are you going to do if there are 3, or 4, or 5 files with the same filename in an item? add 5 underscores?

I would do it differently. I would save the zip with a structure like this:

```
.
├── itemId01
│   └── attachmentId01
│       └── AttachmentFile
├── itemId02
│   └── attachmentId01
│       └── AttachmentFile
└── itemId03
    ├── attachmentId01
    │   └── AttachmentFile
    └── attachmentId02
        └── AttachmentFile
```

In fact, [this is what I do in my bash script](https://github.com/kpiris/bcbbr/commit/aa870103a37e30ea1738cd93ba079376728c68dc) that exports the attachments using the cli.

But hey! it's just an idea!

Thanks!

djsmith85 commented:
@kpiris Thanks for the comments. 

I mean, I think ultimately it's not very convenient to have files with the same name. Thinking of a folder with all the files having the same name, I'd probably want some more metadata likes file size and last changed. 

I think the issue should be solved within the vault and to prevent using the same filename, as other systems do. Or do you maybe have other thoughts on this too? The only reason this currently works is, because on the server the filenames are encrypted and we can only identify them by id. 

I would have to bring my idea up to discussion with @bitwarden/team-vault-dev. 

You are right that restoring a backup could potentially cause different results, if the filenames are being changed on export. I only did this, to prevent data loss and as an okay temporary solution, until a restore with attachments is being built.

djsmith85 committed:
Create a folder for each attachment using its attachmentId as it's name

This is to prevent name-collision for ciphers that have attachments with the same name.

audreyality commented on libs/tools/export/vault-export/vault-export-core/src/services/individual-vault-export.service.ts:
```diff
@@ -117,7 +117,11 @@ export class IndividualVaultExportService
-        cipherFolder.file(attachment.fileName, decBuf);
+
+        // To prevent files with the same name from overwriting each other, a folder with the attachmentId is created and then the attachment is saved within it
+        // This also has the benefit of retaining the attachmentId, which is useful when restoring backups
```
🤔 We shouldn't be trusting the attachment id for anything beyond internal linking within the export. If we restore the fileId on import, it could let one user's files overwrite another user's files.
